### ISSUE TITLE
Chapter 07 - possible error in nLockTime set to "<current block # + 1>" to mitigate fee-sniping

### ISSUE THREAD

**Issue by ctodea** [bitcoinbook/bitcoinbook#997]
Chapter 07 states:

> To achieve this, Bitcoin Core sets the nLocktime on all new transactions to <current block # + 1> and sets the nSequence on all the inputs to 0xFFFFFFFE to enable nLocktime.

Wouldn't setting the `nLockTime` to a block **in the future** (as <current block # + 1> suggests) make the transaction invalid before <current block # + 1> (in the present) and hence nodes won't relay it?

Digging a bit in Bitcoin-Core I found this, which I believe to be the first proposal for this mitigation: https://github.com/bitcoin/bitcoin/pull/2340

Where the nLockTime is set to either:

```
// in the past
txNew.nLockTime = std::max(0, chainActive.Height() - 10); 
```
or randomly to 
```
// also in the past
    if (GetRandInt(10) == 0)
        txNew.nLockTime = std::max(0, (int)txNew.nLockTime - GetRandInt(100));
```
The code has been improved since then (see https://github.com/bitcoin/bitcoin/pull/2340) but the idea remains the same.

Am I missing something or is this an error?

**Comment by danielabrozzoni** [comment]
> Wouldn't setting the nLockTime to a block in the future (as <current block # + 1> suggests) make the transaction invalid before <current block # + 1> (in the present) and hence nodes won't relay it?

Exactly! At the moment Bitcoin Core either sets the nLockTime to <current block#>, or in the past:

https://github.com/bitcoin/bitcoin/blob/45d8b1e94a8e3e68f210d496330f5e16219d2a8d/src/wallet/spend.cpp#L632-L641

I agree that the paragraph should be updated.
